nmosnode/aggregator.py

```diff
@@ -83,58 +80,288 @@ def __init__(self, logger=None, mdns_updater=None):
+        while True:
+            self.aggregator = self._get_aggregator()
+            if self.aggregator is None:
+                self.logger.writeDebug("Failed to find registration API")
+                break
+            self.logger.writeDebug("Aggregator set to: {}".format(self.aggregator))
+
+            if self._heartbeat_operation():
+                # Successfully registered Node
+                break
+
+    def _registered_operation(self):
+        """In Registered operation, the Node is registered so a heartbeat will be performed,
+        if the heartbeat is successful the Node will wait 5 seconds before attempting another heartbeat.
+        Else another aggregator will be selected"""
+        if not self._heartbeat_operation():
+            # Heartbeat failed
+            # Flag to update cached list of aggregators
+            self.aggregator = None
+            self._aggregator_failure = True
+
+    def _heartbeat_operation(self):
+        """Performs a heartbeat to registered aggregator
+        If heartbeat fails it will take actions to correct the error, by re-registering the Node
+        If successfull will return True, else will return False"""
+        if not self.aggregator:
+            return False
+        try:
+            self.logger.writeDebug("Sending heartbeat for Node {}"
+                                   .format(self._node_data["node"]["data"]["id"]))
+            R = self._send("POST", self.aggregator, self.aggregator_apiversion,
+                           "/health/nodes/{}".format(self._node_data["node"]["data"]["id"]))
+
+            if R.status_code == 200 and self._node_data["registered"]:
+                # Continue to registered operation
+                self._registered()
+                heartbeat_wait = 5
+                while heartbeat_wait > 0 and self._running:
+                    gevent.sleep(1)
+                    heartbeat_wait -= 1
+                return True
+
+            elif R.status_code in [200, 409]:
+                # Delete node from registry
+                if self._unregister_node(R.headers.get('Location')):
+                    return self._register_node()
+                else:
+                    # Try next registry
+                    return False
+
+        except InvalidRequest as e:
+            if e.status_code == 404:
+                # Re-register
+                self.logger.writeWarning("404 error on heartbeat. Marking Node for re-registration")
+                self._node_data["registered"] = False
+                return self._register_node()
+            else:
+                # Other error, try next registry
+                return False
+        except ServerSideError:
+            self.logger.writeWarning("Server Side Error on heartbeat. Trying another registry")
+            return False
+        except Exception:
+            # Re-register
+            self.logger.writeWarning("Unexpected error on heartbeat. Marking Node for re-registration")
```

Comment:
I got this error several times when shutting down the service. Whilst it's not a major problem at that time, perhaps the generic exception needs identifying and handling more cleanly in this specific case?

```
2019-11-11 09:37:18,544 WARN  python.facade.aggregator_proxy Unexpected error on heartbeat. Marking Node for re-registration
2019-11-11 09:37:18,545 WARN  python.facade.aggregator_proxy Unexpected error on heartbeat. Marking Node for re-registration
2019-11-11 09:37:18,546 WARN  python.facade.aggregator_proxy Unexpected error on heartbeat. Marking Node for re-registration
2019-11-11 09:37:18,546 WARN  python.facade.aggregator_proxy Unexpected error on heartbeat. Marking Node for re-registration
```

Reply:
Caused by Node Service calling unregister(), which adds DELETE to request queue and instantly sets node_data to None.
Changed to not set node_data to None until resource unregistered request sent
Also heartbeat should not be performed  unless there is node_data